src/controls/OneOfControl.tsx

```diff
@@ -0,0 +1,309 @@
+    setSelectedIndex(combinatorIndex)
+  }, [])
+
+  const prevSelectedIndex = usePreviousValue(selectedIndex)
+  const [dataForPreviousSchemas, setDataForPreviousSchemas] = useState<
+    Record<number, unknown>
+  >({})
+  useEffect(() => {
+    if (
+      selectedIndex !== prevSelectedIndex &&
+      prevSelectedIndex !== undefined
+    ) {
+      setDataForPreviousSchemas({
+        ...dataForPreviousSchemas,
+        [prevSelectedIndex]: data as unknown,
+      })
+      if (dataForPreviousSchemas[selectedIndex]) {
+        handleChange(path, dataForPreviousSchemas[selectedIndex])
+      } else {
+        handleCombinatorTypeChange({
+          handleChange,
+          combinatorIndex: selectedIndex,
+          renderInfos: oneOfRenderInfos,
+          path,
+          rootSchema,
+        })
+      }
+    }
+  }, [
+    data,
+    dataForPreviousSchemas,
+    handleChange,
+    oneOfRenderInfos,
+    path,
+    prevSelectedIndex,
+    rootSchema,
+    selectedIndex,
+  ])
+
+  const computeLabel = (oneOfRenderInfo: CombinatorSubSchemaRenderInfo) => {
+    const label = oneOfRenderInfo.label.startsWith("oneOf")
+      ? oneOfRenderInfo.schema.title
+      : oneOfRenderInfo.label
+    return startCase(label)
+  }
+
+  return (
+    <Space direction="vertical" style={{ width: "100%" }} size="middle">
+      <CombinatorTitle uischema={uischema} schema={schema} />
+      {shouldUseRadioGroupSwitcher(oneOfOptionType) && (
+        <RadioGroupSwitcher
+          radioProps={{
+            style: { marginBottom: "12px" },
+            ...(oneOfOptionType === "button"
+              ? { optionType: "button", buttonStyle: "solid" }
+              : {}),
+          }}
+          options={oneOfRenderInfos.map((oneOfRenderInfo, index) => ({
+            // revert this when this is merged & released: https://github.com/eclipsesource/jsonforms/pull/2165#issuecomment-1640981617
+            label: computeLabel(oneOfRenderInfo),
+            value: index,
+          }))}
+          onChange={handleRadioChange}
+          value={selectedIndex}
+        />
+      )}
+      {oneOfOptionType === "dropdown" && (
+        <Select
+          options={oneOfRenderInfos.map((oneOfRenderInfo, index) => ({
+            // revert this when this is merged & released: https://github.com/eclipsesource/jsonforms/pull/2165#issuecomment-1640981617
+            label: computeLabel(oneOfRenderInfo),
+            value: index,
+          }))}
+          onChange={handleSelectChange}
+          defaultValue={selectedIndex}
+        />
+      )}
+      {oneOfOptionType === "toggle" && (
+        <Space>
+          <Switch
+            onChange={handleSwitchChange}
+            defaultChecked={indexOfFittingSchema === 1}
+          />
+          {appliedUiSchemaOptions.toggleLabel && (
+            <Typography.Text>
+              {appliedUiSchemaOptions.toggleLabel}
+            </Typography.Text>
+          )}
+        </Space>
+      )}
```

Comment:
I'm gonna see if I can put this in a separate component before merging

Reply:
heh I just pushed this refactor =)

Reply:
Yeah, assuming you're talking about CombinatorSchemaSwitcher, 200 lines is a bit long